auth_oidc/tests/test_auth_oidc_logout.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2021 ACSONE SA/NV <https://acsone.eu>
+# License: AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+import contextlib
+import logging
+from urllib.parse import parse_qsl, urljoin, urlparse
+
+from werkzeug.urls import url_encode
+
+import odoo
+from odoo.tests import common
+from odoo.tools.misc import DotDict
+
+from odoo.addons.website.tools import MockRequest
+
+from ..controllers.main import OpenIDLogout
+
+BASE_URL = "http://localhost:%s" % odoo.tools.config["http_port"]
+CLIENT_ID = "auth_oidc-test"
+LOGIN_PATH = "/web/login"
+OIDC_BASE_LOGOUT_URL = "http://keycloak"
+OIDC_LOGOUT_PATH = "/logout"
+
+logger = logging.getLogger(__name__)
+
+
+@contextlib.contextmanager
+def create_request(env, user_id, user_logout_func):
+    with MockRequest(env) as request:
+        request.httprequest.url_root = BASE_URL + "/"
+        request.session = DotDict(uid=user_id, logout=user_logout_func)
+        yield request
+
+
+class TestOpenIDLogout(common.HttpCase):
+    @staticmethod
+    def mock_logout_user(keep_db):
+        logger.info("Logging out user in Odoo")
+
+    def setUp(self):
+        super().setUp()
+        # search our test provider and bind the demo user to it
+        self.provider = self.env["auth.oauth.provider"].search(
+            [("client_id", "=", CLIENT_ID)]
+        )
+        self.assertEqual(len(self.provider), 1)
+
+    def _prepare_login_test_user(self, provider_id):
+        user = self.env.ref("base.user_demo")
+        user.write({"oauth_provider_id": provider_id, "oauth_uid": user.login})
+        return user
+
+    def _set_test_oidc_logout_url(self, end_session_endpoint):
+        self.provider.write({"end_session_endpoint": end_session_endpoint})
+
+    def test_skip_oidc_logout_for_user(self):
+        """Test that oidc logout is skipped if user is not associated to a provider"""
+        user = self._prepare_login_test_user(None)
+        with create_request(self.env, user.id, self.mock_logout_user):
+            resp = OpenIDLogout().logout()
+            self.assertEqual(LOGIN_PATH, resp.location)
+
+    def test_skip_oidc_logout_for_all_users(self):
+        """Test that oidc logout is skipped for all users if provider has no logout url"""
+        self.assertFalse(None, self.provider.end_session_endpoint)
```

Comment:
This assert is useless as None is always falsy, and if you remove it, this test looks identical to the previous one?

Reply:
Good catch it was an oversight.

The tests are not identical, the previous test is for the scenario where a user is not associated to an idp provider, while this test is for the scenario where no end_session_endpoint is configured for the idp provider, the difference is on lines 58 and 66.